File: concepts/python/src/list_nodes.py
from collections.abc import Iterable, Iterator
from typing import Optional, TypeVar, Generic

T = TypeVar('T', bound='ListNode')


class ListNode(Generic[T], Iterable):
    def __init__(self, value: int):
        self.value: int = value
        self.next: T | None = None

    def has_next(self) -> bool:
        return self.next is not None

    def iterate_to(self, position: int) -> 'ListNode':
        node: ListNode = self
        for _ in range(position):
            node = node.next
        return node

    def iterate_last(self) -> 'ListNode':
        node: ListNode = self
        while node.has_next():
            node = node.next
        return node

    def __iter__(self):
        return self.ListNodeIterator(self)

    def __hash__(self):
        return hash(self.value)

    def __eq__(self, other):
        if self is other:
            return True
        if not isinstance(other, ListNode):
            return False
        node1: ListNode = self
        node2: ListNode = other
        while node1 and node2:
            if node1.value != node2.value:
                return False
            node1 = node1.next
            node2 = node2.next
        return node1 == node2

    class ListNodeIterator(Iterator['ListNode']):
        def __init__(self, node: 'ListNode'):
            self.node = node

        def __next__(self):
            if not self.node:
                raise StopIteration
            temp = self.node
            self.node = self.node.next
            return temp


class SinglyListNode(ListNode):
    def __init__(self, value: int = 0, next2: Optional['SinglyListNode'] = None):
        super().__init__(value)
        self.next = next2

    @staticmethod
    def of(*values: int) -> 'SinglyListNode':
        if not values:
            raise ValueError("Can't be empty.")
        head: SinglyListNode = SinglyListNode(values[0])
        next2: SinglyListNode = head
        for val in values[1:]:
            next2.next = SinglyListNode(val)
            next2 = next2.next
        return head


class DoublyListNode(ListNode):
    def __init__(
        self,
        value: int = 0,
        next2: Optional['DoublyListNode'] = None,
        prev2: Optional['DoublyListNode'] = None,
    ):
        super().__init__(value)
        self.next = next2
        self.prev: DoublyListNode | None = prev2


SINGLY_LIST_NODE_ZERO: SinglyListNode = SinglyListNode(0)

File: kotlin/src/main/kotlin/concepts/ListNodes.kt
package concepts

class SinglyListNode : ListNode<SinglyListNode> {
    constructor() : super(0)

    constructor(value: Int) : super(value)

    constructor(value: Int, next: SinglyListNode) : super(value) {
        this.next = next
    }

    companion object {
        val ZERO = SinglyListNode(0)

        fun of(vararg values: Int): SinglyListNode? {
            if (values.isEmpty()) {
                return null
            }
            val head = SinglyListNode(values[0])
            var next = head
            for (i in 1 until values.size) {
                next.next = SinglyListNode(values[i])
                next = next.next!!
            }
            return head
        }
    }
}

class DoublyListNode : ListNode<DoublyListNode> {
    var prev: DoublyListNode? = null

    constructor() : super(0)

    constructor(value: Int) : super(value)

    constructor(value: Int, next: DoublyListNode) : super(value) {
        this.next = next
    }

    constructor(value: Int, next: DoublyListNode, prev: DoublyListNode) : super(value) {
        this.next = next
        this.prev = prev
    }

    companion object {
        fun of(vararg values: Int): DoublyListNode? {
            if (values.isEmpty()) {
                return null
            }
            val head = DoublyListNode(values[0])
            var prev: DoublyListNode? = null
            var next = head
            for (i in 1 until values.size) {
                val temp = next

                next.prev = prev
                next.next = DoublyListNode(values[i])
                next = next.next!!

                prev = temp
            }
            return head
        }
    }
}

open class ListNode<T : ListNode<T>>(var value: Int) : Iterable<T> {
    var next: T? = null

    fun hasNext(): Boolean = next != null

    override fun iterator(): Iterator<T> = ListNodeIterator(this as T)

    override fun equals(other: Any?): Boolean {
        if (other !is ListNode<*>) {
            return false
        }
        var node1: ListNode<T>? = this
        var node2: ListNode<T>? = other as? ListNode<T>
        while (node1 != null && node2 != null) {
            if (node1.value != node2.value) {
                return false
            }
            node1 = node1.next
            node2 = node2.next
        }
        return node1 == null && node2 == null
    }

    override fun hashCode(): Int = value.hashCode()

    private inner class ListNodeIterator(private var current: T?) : Iterator<T> {
        override fun hasNext(): Boolean = current != null

        override fun next(): T {
            val temp = current!!
            current = current?.next
            return temp
        }
    }
}
